tools/clang/lib/AST/HlslTypes.cpp

```diff
@@ -609,7 +604,8 @@ bool IsUserDefinedRecordType(clang::QualType QT) {
   const clang::Type *Ty = QT.getCanonicalType().getTypePtr();
   if (const RecordType *RT = dyn_cast<RecordType>(Ty)) {
     const RecordDecl *RD = RT->getDecl();
-    if (RD->isImplicit())
+    // Built-in ray tracing struct types are considered user defined types.
+    if (RD->isImplicit() && !IsHLSLBuiltinRayAttributeStruct(QT))
```

Comment:
This does seem to treat raytracing structs as UDTs and was what I was made to expect from this change.

Reply:
This was the actual key change which fixed the issue.  After this change, `IsHLSLNumericUserDefinedType` and `IsHLSLCopyableAnnotatableRecord` would have returned identical results (no point in calling `IsHLSLBuiltinRayAttributeStruct` there), so I collapsed them into one function under the name `IsHLSLCopyableAnnotatableRecord`.